// @vonage/server-sdk/lib/Channel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _index = _interopRequireDefault(require("./index"));

var _fs = _interopRequireDefault(require("fs"));

var _querystring = _interopRequireDefault(require("querystring"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Channel {
  static get PATH() {
    return "/v0.1/messages";
  }

  constructor(credentials, options) {
    this.creds = credentials;
    this.options = options; // Used to facilitate testing of the call to the underlying object

    this._nexmo = this.options.nexmoOverride || _index.default;

    this._nexmo.initialize(this.creds.apiKey, this.creds.apiSecret, this.options);
  }

  send(to, from, message, callback) {
    var params = {
      to: to,
      from: from,
      message: message
    };
    return this.options.api.post(Channel.PATH, params, callback, true, {
      "Content-Type": "application/json"
    });
  }

}

var _default = Channel;
exports.default = _default;
module.exports = exports.default;
