// @vonage/server-sdk/lib/Number.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _index = _interopRequireDefault(require("./index"));

var _Pricing = _interopRequireDefault(require("./Pricing"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Number {
  /**
   * @param {Credentials} credentials
   *    credentials to be used when interacting with the API.
   * @param {Object} options
   *    Addition Number options.
   */
  constructor(credentials) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    this.creds = credentials;
    this.options = options;
    this._pricing = new _Pricing.default(credentials, options); // Used to facilitate testing of the call to the underlying object

    this._vonage = this.options.vonageOverride || _index.default;

    this._vonage.initialize(this.creds.apiKey, this.creds.apiSecret, this.options);
  }
  /**
   * TODO: remove with next major release
   */


  getPricing() {
    this._pricing.get.apply(this, arguments);
  }
  /**
   * TODO: remove with next major release
   */


  getPhonePricing() {
    this._pricing.getPhone.apply(this, arguments);
  }
  /**
   * TODO: document
   */


  get() {
    this._vonage.getNumbers.apply(this._vonage, arguments);
  }
  /**
   * TODO: document
   */


  search() {
    this._vonage.searchNumbers.apply(this._vonage, arguments);
  }
  /**
   * TODO: document
   */


  buy() {
    this._vonage.buyNumber.apply(this._vonage, arguments);
  }
  /**
   * TODO: document
   */


  cancel() {
    this._vonage.cancelNumber.apply(this._vonage, arguments);
  }
  /**
   * TODO: document
   */


  update() {
    this._vonage.updateNumber.apply(this._vonage, arguments);
  }

}

var _default = Number;
exports.default = _default;
module.exports = exports.default;
