// @vonage/server-sdk/lib/Pricing.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Utils = _interopRequireDefault(require("./Utils"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Pricing {
  static get PATH() {
    return "/account/{endpoint}/outbound/{type}";
  }

  constructor(credentials, options) {
    this.creds = credentials;
    this.options = options;
  }

  get(type, country, callback) {
    return this.options.rest.get(Pricing.PATH.replace("{endpoint}", "get-pricing").replace("{type}", type), {
      country
    }, callback);
  }

  getFull(type, callback) {
    return this.options.rest.get(Pricing.PATH.replace("{endpoint}", "get-full-pricing").replace("{type}", type), callback);
  }

  getPrefix(type, prefix, callback) {
    return this.options.rest.get(Pricing.PATH.replace("{endpoint}", "get-prefix-pricing").replace("{type}", type), {
      prefix
    }, callback);
  }

  getPhone(type, phone, callback) {
    return this.options.rest.get(Pricing.PATH.replace("{endpoint}", "get-phone-pricing").replace("{type}", type), {
      phone
    }, callback);
  }

}

var _default = Pricing;
exports.default = _default;
module.exports = exports.default;
